Observation analytics

mauberti-bc opened the pull request:
## Links to Jira Tickets
- N/A

## Description of Changes

- Adds an "Analytics" tab on the Survey page above the Observations data table
- Analytics endpoint for calculating summary statistics of observations for multiple Surveys
- Haven't written any unit tests in case the backend changes after review

mauberti-bc committed:
wip: observation analytics

mauberti-bc committed:
analytics table for observations

mauberti-bc committed:
fetch crittberase column names for analytics table

mauberti-bc committed:
include measurements as jsonb in analytics service

mauberti-bc committed:
move critterbase requests to backend

mauberti-bc committed:
cache analytics request for 3 minutes

mauberti-bc committed:
make fix

mauberti-bc committed:
merge latest

mauberti-bc committed:
rename columns in analytics data grid

mauberti-bc committed:
format analytics table

mauberti-bc committed:
merge latest

mauberti-bc committed:
merge dev

MacDeluca commented on api/src/paths/analytics/observations.ts:
```diff
@@ -0,0 +1,167 @@
+      throw new HTTP400('Invalid or missing groupByColumns');
+    }
+
+    const connection = getDBConnection(req['keycloak_token']);
```
You can use req.keycloak_token now

MacDeluca commented on api/src/paths/project/{projectId}/survey/{surveyId}/observations/measurements/index.ts:
```diff
@@ -0,0 +1,134 @@
+
+    defaultLog.debug({ label: 'getSurveyObservationMeasurement', surveyId });
+
+    const connection = getDBConnection(req['keycloak_token']);
```
req.keycloak_token

mauberti-bc committed:
Merge branch 'dev' into observation-analytics

MacDeluca commented on api/src/paths/analytics/observations.ts:
```diff
@@ -0,0 +1,167 @@
+  return async (req, res) => {
+    const { surveyIds, groupByColumns, groupByQuantitativeMeasurements, groupByQualitativeMeasurements } = req.query;
+
+    if (!surveyIds || !Array.isArray(surveyIds) || surveyIds?.length === 0) {
```
You can handle this logic in the openapi spec

mauberti-bc committed:
Merge branch 'dev' into observation-analytics

mauberti-bc committed:
update service code

mauberti-bc committed:
remove param checks from endpoint

mauberti-bc committed:
sonarcloud issues

MacDeluca commented on api/src/models/observation-analytics.ts:
```diff
@@ -0,0 +1,67 @@
+  row_count: number;
+  individual_count: number;
+  individual_percentage: number;
+  [key: string]: any; // To accommodate dynamic columns
```
Is this truly ```any``` or something like ```string | number | null | undefined"```

NickPhura replied:
I was able to turn this interface into a zod schema by adding `catchall` (https://zod.dev/?id=catchall) to it to handle the dynamic fields. I had it set to string or number or null, but the upstream type handling blew up, so I had to set it to any.

There was some talk only about this issue, but no solution yet.

MacDeluca commented on api/src/services/analytics-service.ts:
Missing JSDoc for methods

NickPhura replied:
Added

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into observation-analytics

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into observation-analytics

NickPhura committed:
Tweak analytics api hook code to use built-in serializer.
Tweak analytics service - add JSdoc and adjust some functions
Add some unit tests.

NickPhura committed:
Add more zod schema support.
Tweak openapi schema.
Add generated id to analytics response items.
Update frontend table to leverage column visibility to show/hide group-by columns.
Add/Update unit tests.

NickPhura reviewed (dismissed):
🐦

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into observation-analytics

NickPhura reviewed (dismissed):
🐦 🐦

NickPhura committed:
Fix css

NickPhura reviewed (approved):
🐦 🐦 🐦

mauberti-bc committed:
Merge branch 'dev' into observation-analytics

NickPhura committed:
ignore-skip

NickPhura committed:
Merge branch 'observation-analytics' of https://github.com/bcgov/biohubbc into observation-analytics

NickPhura committed:
ignore-skip

mauberti-bc committed:
Merge branch 'dev' into observation-analytics
